carma_wm/src/CARMAWorldModel.cpp

```diff
@@ -450,4 +447,156 @@ CARMAWorldModel::toRoadwayObstacle(const cav_msgs::ExternalObject& object) const
 
   return obs;
 }
+
+void CARMAWorldModel::setRoadwayObjects(const std::vector<cav_msgs::RoadwayObstacle>& rw_objs)
+{
+  roadway_objects_ = rw_objs;
+  return;
+}
+
+std::vector<cav_msgs::RoadwayObstacle> CARMAWorldModel::getRoadwayObjects() const
+{
+  return roadway_objects_;
+}
+
+std::vector<cav_msgs::RoadwayObstacle> CARMAWorldModel::getInLaneObjects(const lanelet::ConstLanelet& lanelet) const
+{
+  // Check if the map is loaded yet
+  if (!semantic_map_ || semantic_map_->laneletLayer.size() == 0)
+  {
+    throw std::invalid_argument("Map is not set or does not contain lanelets");
+  }
+
+  // Check if the lanelet is in map
+  if (semantic_map_->laneletLayer.find(lanelet.id()) == semantic_map_->laneletLayer.end())
+  {
+    throw std::invalid_argument("Lanelet is not on the map");
+  }
+
+  // Check if any roadway object is registered
+  if (roadway_objects_.size() == 0)
+  {
+    return std::vector<cav_msgs::RoadwayObstacle>{};
+  }
+
+  // Get all lanelets on current lane
+  std::vector<lanelet::ConstLanelet> lane = map_routing_graph_->previous(lanelet, false);
+  lane.push_back(lanelet);
+  std::vector<lanelet::ConstLanelet> remaining_lane = map_routing_graph_->following(lanelet, false);
+  lane.insert(lane.end(), remaining_lane.begin(), remaining_lane.end());
+
+  // Initialize useful variables
+  std::vector<cav_msgs::RoadwayObstacle> lane_objects, roadway_objects_copy = roadway_objects_; 
+  
+  /*
+  * Get all in lane objects
+  * For each lane, we check if each object is on it. 
+  * Complexity N*K, where N: num of lanelets, K: num of objects
+  */
+  int curr_idx;
+  std::queue <int> obj_idxs_queue;
+
+  // Create an index queue for roadway objects to quickly pop the idx if associated 
+  // lanelet is found. This is to reduce number of objects to check as we check new lanelets
+  for (int i = 0; i < roadway_objects_copy.size(); i++)
+  {
+    obj_idxs_queue.push(i);
+  }
+  // check each lanelets
+  for (auto lanelet: lane)
+  {
+    int checked_queue_items = 0;
+    // check each objects
+    while (checked_queue_items < obj_idxs_queue.size())
```

Comment:
Rather than checking the intersection of every object with every lanelet. I feel like you could just leverage the already computed lanelet_id and just check if that lanelet is in your id set.